src/netif.rs

```diff
@@ -873,20 +873,22 @@ where
 #[cfg(feature = "alloc")]
 mod driver {
     use core::borrow::Borrow;
+    use core::marker::PhantomData;
 
     use crate::handle::RawHandle;
     use crate::sys::*;
 
     use super::EspNetif;
 
-    pub struct EspNetifDriver<T>
+    pub struct EspNetifDriver<'d, T>
```

Comment:
I don't mind introducing the `'d` lifetime (in fact, I was planning to do it anyway, so good catch!), as it would allow you to borrow non-static stuff into the `tx` closure, BUT you have to realize this is **unsafe in Rust**. And then reason why Rust allows you to do it, is because the code inside `EspNetifDriver` is littered with `unsafe` so Rust can't notice the unsafeness of what you are doing.

It is unsafe because, the moment you call `core::mem::forget(your-esp-netif-driver)`, from the point of view of Rust - everything you have borrowed in the `tx` closure is no longer borrowed and can be re-used.

Except the underlying C code **does not think that's the case and will happily continue to use your borrowed stuff (including the closure itself) thus leading to memory corruption**. That's exactly what Rust can't notice - it can't "see through" the C code and realize that our non-static closure is in fact being sent to a hidden FreeRTOS task/thread which will _continue to operate even if you `core::mem::forget` the driver_.

What we do in [other places of the code](https://github.com/esp-rs/esp-idf-svc/blob/master/src/wifi.rs#L1004) is to rename the `new_*` constructors that take non-static closures to `new_nonstatic_*`, slap a big comment that users should be aware that these are unsafe (as in, users should not use `core::mem::forget` on a driver created with a non-static constructor), and well - mark them as `unsafe`.

So I suggest you rename the eixsting constructors to `new_nonstatic_*` slap the same comment on top, and then re-implement regular `new_*` constructors that take `'static` closures and which call into the `new_nonstatic_*` ones (as they become safe when `'d: 'static`).

Reply:
The reason I introduced the `'d` lifetime is because the serial device (in this case a `UartTxDriver`) that is given to the `tx` closure is not a `'static` lifetime.